finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala

```diff
@@ -0,0 +1,182 @@
+private[netty3] class HttpProxyConnectHandler(
+                                              host: String,
+                                              credentialsOption: Option[Transporter.Credentials],
+                                              httpClientCodec: ChannelHandler = new HttpClientCodec()) // exposed for testing
+  extends SimpleChannelHandler
+{ self =>
+
+  private[this] val httpCodecKey: String = "http proxy client codec"
+
+  private[this] val connectFuture = new AtomicReference[ChannelFuture](null)
+
+  private[this] def proxyAuthorizationHeader(c: Credentials): String = {
+    val bytes = "%s:%s".format(c.username, c.password).getBytes(Charsets.Utf8)
+    "Basic " + Base64StringEncoder.encode(bytes)
+  }
+
+  private[this] def fail(ctx: ChannelHandlerContext, t: SocketAddress => Throwable): Unit = {
+    // We "try" because it might be already cancelled and we don't need to handle
+    // cancellations here - it's already done by `proxyCancellationsTo`.
+    connectFuture.get().setFailure(t(ctx.getChannel.getRemoteAddress))
+
+  }
+
+  private[this] def fail(c: Channel, t: SocketAddress => Throwable) {
+    Option(connectFuture.get) foreach { _.setFailure(t(c.getRemoteAddress)) }
+    Channels.close(c)
+  }
+
+  override def connectRequested(ctx: ChannelHandlerContext, e: ChannelStateEvent): Unit = {
+    println("entering connectRequested")
+    e match {
+      case de: DownstreamChannelStateEvent =>
+        if (!connectFuture.compareAndSet(null, e.getFuture)) {
+          fail(ctx.getChannel, (remoteAddress : SocketAddress) => new InconsistentStateException(remoteAddress))
+          return
+        }
+
+        // proxy cancellation
+        val wrappedConnectFuture = Channels.future(de.getChannel, true)
+        de.getFuture.addListener(new ChannelFutureListener {
+          def operationComplete(f: ChannelFuture) {
+            if (f.isCancelled)
+              wrappedConnectFuture.cancel()
+          }
+        })
+
+        // Proxy failures here so that if the connect fails, it is
+        // propagated to the listener, not just on the channel.
+        wrappedConnectFuture.addListener(new ChannelFutureListener {
+          def operationComplete(f: ChannelFuture) {
+            if (f.isSuccess || f.isCancelled) {
+              return
+            }
+
+            fail(f.getChannel, (_: SocketAddress) => f.getCause)
+          }
+        })
+
+        // We propagate the pipeline with a new promise thereby delaying the original connect's
+        // satisfaction.
+        // TODO: Figure out why the remoteAddress is
+        val wrappedEvent = new DownstreamChannelStateEvent(
+          de.getChannel, wrappedConnectFuture,
+          de.getState, de.getValue)
+
+
+        println(s"Connecting to ${wrappedEvent.getChannel.getRemoteAddress} with $wrappedEvent")
+        super.connectRequested(ctx, wrappedEvent)
+    }
+
+  }
+
+  override def writeRequested(ctx: ChannelHandlerContext, e: MessageEvent): Unit = {
```

Comment:
Let's not override it. I think we're ok assuming that nothing is writing into the pipeline on N3 until the connect future is satisfied.